src/flint/types/nmod_poly.pyx

```diff
@@ -191,6 +198,115 @@ cdef class nmod_poly(flint_poly):
     def is_one(self):
         return <bint>nmod_poly_is_one(self.val)
 
+    def is_gen(self):
+        return <bint>nmod_poly_is_gen(self.val)
+
+    def reverse(self, degree=None):
+        """
+        Return a polynomial with the coefficients of this polynomial
+        reversed.
+
+        If ``degree`` is not None, the output polynomial will be zero-padded
+        or truncated before being reversed. NOTE: degree must be non-negative.
+
+            >>> f = nmod_poly([1,2,3,4,5], 163)
+            >>> f.reverse()
+            x^4 + 2*x^3 + 3*x^2 + 4*x + 5
+            >>> f.reverse(degree=1)
+            x + 2
+            >>> f.reverse(degree=100)
+            x^100 + 2*x^99 + 3*x^98 + 4*x^97 + 5*x^96
+        """
+        cdef nmod_poly res
+        cdef slong d
+        
+        if degree is not None:
+            d = degree
+            if d != degree or d < 0:
+                raise ValueError(f"degree argument must be a non-negative integer, got {degree}")
+            length = d + 1
+        else:
+            length = nmod_poly_length(self.val)
+
+        res = nmod_poly.__new__(nmod_poly)
+        nmod_poly_init_preinv(res.val, self.val.mod.n, self.val.mod.ninv)
+        nmod_poly_reverse(res.val, self.val, length)
+        return res
+
+    def inverse_series_trunc(self, slong n):
+        """
+        Returns the inverse of ``self`` modulo `x^n`.
+
+            >>> f = nmod_poly([123, 129, 63, 14, 51, 76, 133], 163)
+            >>> f.inverse_series_trunc(3)
+            159*x^2 + 151*x + 110
+            >>> f.inverse_series_trunc(4)
+            23*x^3 + 159*x^2 + 151*x + 110
+            >>> f.inverse_series_trunc(5)
+            45*x^4 + 23*x^3 + 159*x^2 + 151*x + 110
+        """
+        cdef nmod_poly res
+        res = nmod_poly.__new__(nmod_poly)
+        nmod_poly_init_preinv(res.val, self.val.mod.n, self.val.mod.ninv)
+        nmod_poly_inv_series(res.val, self.val, n)
+        return res
```

Comment:
More checking is needed here:
```python
In [3]: flint.nmod_poly(2, 3).inverse_series_trunc(-3)
Segmentation fault (core dumped)

In [17]: flint.nmod_poly(0, 3).inverse_series_trunc(3)
Flint exception (General error):
    Exception (nmod_poly_inv_series). Division by zero.
Aborted (core dumped)
```

Reply:
Urgh, im tempted to remove the commit for nmod_poly and simply not touch the class until I have time to write a proper context object to do the necessary checks

Reply:
From the docs we have:

> Given Q find Qinv such that Q * Qinv is 1 modulo . The constant coefficient of Q must be invertible modulo the modulus of Q. An exception is raised if this is not the case or if n = 0. This function can be viewed as inverting a power series.

So I guess we need something like

```
check = self.leading_coefficient().is_invertible()
```

Also we need `n != 0`. In fact, does allowing `n < 1` make sense in this context? Why does flint use an `slong` instead of a `ulong`...

Reply:
i dont think we currently have the set up to fix worrying about the bad inversion here... It's something I noticed before #73 and #124 -- i think the whole class needs to have a huge refactor

Reply:
As long as it works for prime modulus that is okay.

I think we just need to check here that `self` is not zero and `n` is not negative and also make it clear in code comment and/or docstring that the modulus needs to be prime.

In SymPy I have now disabled the use of python-flint for non-prime modulus because there are too many operations that fail. As you say a context is needed.